### DISCUSSION TITLE
Performance issue on regex and incorrect working of limitations (windows)

### DISCUSSION THREAD

**Discussion by GeorgeAlexandria** [cvc5/cvc5#7462] [Q&A]
Hello,

I'm used **cvc4 v1.8** (besides, behavior is reproduced on the **cvc4 v1.7**, presumably also on more old versions) and try to solve a few assertions (that were simplified previously) with restrictions as regular expressions:

```
(set-logic ALL_SUPPORTED)
(push)
(declare-const id1 String)
(assert (str.in.re id1 (re.+ (re.union (re.range "\x20" "\x7e") (str.to.re "\x0A") (str.to.re "\x0D")) )))
(assert (not (str.contains id1 "fonts")))
(check-sat)
(pop)
```

Firstly, when cvc4 is run with `-L smt2.5 --output-lang smt2.5 --produce-models --strings-exp` arguments it takes more than 9GB of RAM at a few minutes and doesn't seem, that it will stop at any rational time. I assume that it is a performance and resource issue.

Secondly, when I try to add time limitations `-L smt2.5 --output-lang smt2.5 --produce-models --strings-exp --tlimit-per 1000` (or `--tlimit 1000`) cvc4 still try to satisfy assertions within a few minutes and doesn't stop. I assume that it is a issue on working of limitations.

Finally, I try to add resource limitations by `--rlimit[-per]` passing 100, 1000, 10000, but seems that I don't understand what does it limite? Is it a RAM memory in kb, mb or gb or is it some count of inner instruction? So, this limitation also doesn't help me – cvc4 still continue working a few minutes on `--rlimit-per 10000` and immediately stop working on `--rlimit-per 1000`.

By the way, at a first times I passed `--incremental` flag with arguments and stopped pass them when found [this issue](https://github.com/CVC4/cvc4-projects/issues/202)

If I miss something, please let me know about it.

**Comment by ajreynol** [comment]
Internal note: this benchmark is solved "sat" quickly on master.

@nafur can you explain more about the usage of rlimit?

**Comment by nafur** [comment]
Up until 1.8, the internal resource limitation (in particular tlimit) were somewhat quirky and sometimes just did not work. It has undergone a major refactoring and should work better now.
As for rlimit, the most important thing is that `--rlimit` does not limit the amount of memory, but the number of resources spent during computation in a more abstract sense: certain events (whose frequency tends to be an indicator for runtime) "spend resources" and rlimit limits the overall amount of resources that is available. Events that spend resources are, for example, sat conflicts, a simplex pivoting step, a theory check or a sat restart. Mostly, it is a way to limit the runtime in a less precise, but deterministic way.

**Comment by GeorgeAlexandria** [comment]
@ajreynol, do you have some nightly build of current master or do I should build master by myself to check the current behavior on master and compare with v1.8?

@nafur, is it a way to understand how many events (or average of them) would be produced during computation assert to propose how much value should be used for `--rlimit`? If it cannot be done seems that this limit unusable for external users that do not know how this events are raised and how much them..

**Comment by ajreynol** [comment]
Hi, you can download the latest master binary here: https://cvc4.github.io/downloads.html.

**Comment by nafur** [comment]
The resource counting behind `--rlimit` is not meant to understand what the solver is doing. The main use case is the following: assume you have regressions for a tool that uses cvc4 as a back end solver. For regressions, you usually do not want to impose time limits, because timings can be influences by all kinds of things. `--rlimit` gives you a way to limit the runtime of the solver by not actually limiting the time, but a deterministic measure. So usually you run the solver, figure out how much of these resources were spent, and then use this value as limit for your regressions.

**Comment by GeorgeAlexandria** [comment]
@nafur, understand this usages, but it not so useful for external usages by my opinion, because the main goal that you do not know (and cannot obtain) how much resources were spent by computation, because you (external user) do not know how them are used and the counter of taking resources behind of `--rlimit` doesn't exist. User just can obtain from the many retrying that the asserts will pass by the some values of `--rlimit` or fail by another values and this values may change in the new versions and you cannot tell that is bad or not, because you cannot "touch" this abstract resource to say that. If the new version spent more this resource but take the same time, memory, threads and any other "touch" resource is it bad or not? I think it is dependent.
Anyway, thanks a lot for explanation, I understood the main approach for introduction `--rlimit`.

@ajreynol, thanks for the link, I will try the current build in a several days on the current assertions and on the other non simplified examples and will close or update issue according the results.

**Comment by nafur** [comment]
You can obtain the number of resources spent by the solver by passing `--stats` and looking for `resource::resourceUnitsUsed`.

You are right in some sense, that the use case for this feature is somewhat rare and most users do not need it. Rest assured, though, that we have users that want this exact feature.

**Comment by GeorgeAlexandria** [comment]
@ajreynol, I tried nightly build from [2021-02-09](http://cvc4.cs.stanford.edu/downloads/builds/win64-opt/unstable/cvc4-2021-02-09-win64-opt.exe) with assertion and launch args from above and all issues still persist on it. Did I miss something?

**Comment by ajreynol** [comment]
I just downloaded the (linux) binary from 2021-02-09, adding --incremental in addition to the options you specify on the benchmark make the benchmark solved "sat" quickly.  Which exact binary (what OS) are you using?

**Comment by GeorgeAlexandria** [comment]
I tried under windows. Build was taken from here: http://cvc4.cs.stanford.edu/downloads/builds/win64-opt/unstable/
